src/wp-settings.php

```diff
@@ -412,6 +412,124 @@
+$_found_incompatible_for_core_plugins = array();
+
+/**
+ * Checks if the given plugin is compatible with this WordPress version.
+ *
+ * @since 6.5.0
+ *
+ * @global $_wp_for_core_plugins_compat_version
+ * @global $_found_incompatible_for_core_plugins
+ *
+ * @param string $plugin_absolute_path Absolute path to the plugin's file.
+ */
+$_is_plugin_compatible_with_wp = static function ( $plugin_absolute_path ) use ( $_wp_for_core_plugins_compat_version ) {
+	global $_found_incompatible_for_core_plugins;
+
+	static $plugins_dir_strlen;
+	if ( empty( $plugins_dir_strlen ) ) {
+		$plugins_dir_strlen = strlen( WP_PLUGIN_DIR . '/' );
+	}
+
+	$plugin = substr( $plugin_absolute_path, $plugins_dir_strlen );
+
+	// Not a for-core plugin.
+	$min_compat_version = _get_plugin_wp_min_compatible_version( $plugin );
+	if ( ! $min_compat_version ) {
+		return true;
+	}
+
+	// The plugin is already marked as incompatible.
+	if ( isset( $_found_incompatible_for_core_plugins[ $plugin ] ) ) {
+		return false;
+	}
+
+	// Get the Name and Version from the plugin's header.
+	require_once ABSPATH . 'wp-admin/includes/plugin.php';
+	$plugin_data = get_plugin_data(
+		$plugin_absolute_path,
+		array(
+			'Name'    => 'Plugin Name',
+			'Version' => 'Version',
+		),
+		'plugin'
+	);
+
+	// Whoops, something went wrong. Bail out.
+	if ( ! ( isset( $plugin_data['Version'] ) && isset( $plugin_data['Name'] ) ) ) {
+		return true;
+	}
+
+	// Plugin is compatible.
+	if ( version_compare( $plugin_data['Version'], $min_compat_version, '>=' ) ) {
+		return true;
+	}
+
+	// Found an incompatible for-core plugin. Add it to the found global for later batch processing.
+	$_found_incompatible_for_core_plugins[ $plugin ] = array(
+		'plugin_absolute_path' => $plugin_absolute_path,
+		'plugin_name'          => $plugin_data['Name'],
+		'version_deactivated'  => $plugin_data['Version'],
+		'version_compatible'   => $min_compat_version,
+	);
+
+	return false;
+};
+
+/**
+ * Handle incompatible for-core plugins.
+ *
+ * The handler does the following:
+ *     * Removes incompatible plugins from the "active_plugins" option.
+ *     * Stores incompatible plugins in "wp_force_deactivation_incompatible_plugins" option, which is used
+ *       in wp-admin to notify the user.
+ *
+ * @since 6.5.0
+ *
+ * @global $_found_incompatible_for_core_plugins
+ */
+$_handle_incompatible_for_core_plugins = static function () {
+	global $_found_incompatible_for_core_plugins;
+
+	$active_plugins           = (array) get_option( 'active_plugins', array() );
+	$active_plugins_by_plugin = array_flip( $active_plugins );
+
+	// Remove each of the found incompatible plugins from the "active_plugins" option.
+	foreach ( $_found_incompatible_for_core_plugins as $plugin_slug => $plugin ) {
+		unset( $active_plugins[ $active_plugins_by_plugin[ $plugin_slug ] ] );
+	}
+
+	// Update the 'active plugins' option, which no longer includes the incompatible plugins.
+	update_option( 'active_plugins', $active_plugins );
```

Comment:
What can be done to get rid of this database write during the early bootstrapping?

Ref: performance thoughts here https://github.com/WordPress/wordpress-develop/pull/5855#discussion_r1481570532

Hmm ..

Reply:
Whoops, also need to consider networked plugins, i.e. `'active_sitewide_plugins'`.